src/file_path.h

```diff
@@ -0,0 +1,123 @@
+// Copyright 2021 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NINJA_FILE_PATH_H_
+#define NINJA_FILE_PATH_H_
+
+#include <string>
+
+#ifndef TEXT
+#ifdef  UNICODE
+#define __TEXT(quote) L##quote
+#else   /* UNICODE */
+#define __TEXT(quote) quote
+#endif /* UNICODE */
+#define TEXT(quote) __TEXT(quote)
+#endif
+
+#ifndef _TCHAR_DEFINED
+#ifdef UNICODE
+typedef wchar_t TCHAR, *PTCHAR;
+#else
+typedef char TCHAR, *PTCHAR;
+#endif
+#define _TCHAR_DEFINED
+#endif /* !_TCHAR_DEFINED */
+
+#ifdef UNICODE
+typedef std::wstring file_string_t;
+#else
+typedef std::string file_string_t;
+#endif
+
+inline std::string NarrowPath(const std::wstring& path) {
+  return std::string(path.begin(), path.end());
```

Comment:
Are you sure that this is what you want here? losing information silently doesn't sound like a great idea.

Reply:
The goal of this PR is solely to add long path support by using the unicode version of all windows APIs. It does not actually add support for Unicode, which is a much broader change. So the behavior of what contents a path may contain are the same in this PR as in the existing ninja releases

Reply:
This should at a minimum call WideCharToMultiByte() to ensure that the result matches either UTF-8 or the current codepage, instead of truncating whatever UTF-16 values in the input path to 8-bits. For example, there is one example of calling FindFirstFileExW() explicitly, whose result is passed to this NarrowPath() function, and this is not correct.

https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-widechartomultibyte

Reply:
How would you recommend handling errors from WideCharToMultiByte?

Reply:
It depends on the call site (e.g. for FindFirstFileExW() you would just ignore the file). It's probably better not to have a generic NarrowPath() function at all, and handle the wide-to-byte conversion errors explicitly at the small number of call sites where they could happen.